File: src/llamacpp_perfkit/tui/state.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class BuildInfoView:
    commit_short: str
    branch: str
    backend: str


@dataclass
class ProgressState:
    servers_completed: int = 0
    servers_total: int = 0
    jobs_completed: int = 0
    jobs_total: int = 0
    current_prompt_index: int = 0
    current_prompt_total: int = 0


@dataclass
class CurrentServerView:
    id: str
    context_size: int
    kv_type: str
    n_cpu_moe: int
    spec_type: str
    batch_size: int | None = None
    ubatch_size: int | None = None


@dataclass
class PromptJobView:
    profile: str
    status: str
    phase: str
    duration_seconds: float | None = None
    gen_tok_s: float | None = None
    prompt_tok_s: float | None = None
    min_vram_mib: float | None = None


@dataclass
class BenchmarkTUIState:
    run_id: str
    build_info: BuildInfoView
    model_name: str
    progress: ProgressState = field(default_factory=ProgressState)
    elapsed_seconds: float = 0.0
    eta_seconds: float = 0.0
    lifecycle_state: str = "planning"
    status_message: str = ""
    active_prompt_profile: str | None = None
    current_server: CurrentServerView | None = None
    prompt_jobs: list[PromptJobView] = field(default_factory=list)

File: internal/tui/views/state.go
package views

type BuildInfoView struct {
	CommitShort string
	Branch      string
	Backend     string
}

type ProgressState struct {
	ServersCompleted   int
	ServersTotal       int
	JobsCompleted      int
	JobsTotal          int
	CurrentPrompt      int
	CurrentPromptTotal int
}

type CurrentServerView struct {
	ID          string
	ContextSize int
	KVType      string
	NCPUMOE     int
	SpecType    string
	BatchSize   int
	UBatchSize  int
}

type PromptJobView struct {
	Profile         string
	Status          string
	Phase           string
	DurationSeconds *float64
	GenTokS         *float64
	PromptTokS      *float64
	MinVRAMMiB      *float64
}

type BenchmarkTUIState struct {
	RunID               string
	BuildInfo           BuildInfoView
	ModelName           string
	Progress            ProgressState
	ElapsedSeconds      float64
	ETASeconds          float64
	LifecycleState      string
	StatusMessage       string
	ActivePromptProfile string
	CurrentServer       *CurrentServerView
	PromptJobs          []PromptJobView
}

type StateUpdate struct {
	Apply func(*BenchmarkTUIState)
}

func (s *BenchmarkTUIState) UpsertPrompt(job PromptJobView) {
	for i := range s.PromptJobs {
		if s.PromptJobs[i].Profile == job.Profile {
			if job.Status != "" {
				s.PromptJobs[i].Status = job.Status
			}
			if job.Phase != "" {
				s.PromptJobs[i].Phase = job.Phase
			}
			if job.DurationSeconds != nil {
				s.PromptJobs[i].DurationSeconds = job.DurationSeconds
			}
			if job.GenTokS != nil {
				s.PromptJobs[i].GenTokS = job.GenTokS
			}
			if job.PromptTokS != nil {
				s.PromptJobs[i].PromptTokS = job.PromptTokS
			}
			if job.MinVRAMMiB != nil {
				s.PromptJobs[i].MinVRAMMiB = job.MinVRAMMiB
			}
			return
		}
	}
	s.PromptJobs = append(s.PromptJobs, job)
}
